designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+The solution to both of them is rather simple from a user perspective - the new
+`array[]` syntax can be automatically produced by our pretty-printer, and it is
+easy to rename a variable, but this must be clearly and consistently
+communicated to our users.
+
+Finally, this process provides an opportunity for feedback on changes. The
+deprecation period prior to removal is not only to let users know of our
+intentions, but also to let them tell us if our intentions need to be rethought.
+By communicating the changes and putting deadlines on them, we create a sort of
+community review period during which we can recieve feedback and course-correct
+if we have made a poor choice.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Deprecated minor features may be removed from Stan during minor version changes.
+Here, 'minor change' is restricted to be changes which can be done automatically
+through the use of the stanc3 canonicalizer, or a user-driven find and replace
+operation. By their very nature, minor changes are primarily syntactic. This is
+in contrast to 'major changes', which are by their nature semantic, and require
+human-in-the-loop reasoning to accommodate. Major changes, such the interface
+changes to the differential equation solvers, may be announced via deprecation
+warnings in minor versions, but will not be implemented/removed prior to a
+version 3.0.0.
+
+A minor feature must be deprecated for a minimum of 3 minor versions
+(approximately 1 year on the current release schedule) before removal. The exact
+timeline for a feature's removal will be decided during the development which
+would mark it as deprecated.
+
+A minor feature is deprecated and a candidate for future removal when:
+1. stanc emits a compile-time warning with the following 3 parts:
+    1. A description of the deprecated feature, such as "Use of <- for assignment
+       is deprecated".
+    2. An _action_ the user should take, such as "Use = instead". If stanc3 can
+       automate this for them, we should call that out, leading to a full
+       message like "Use = instead or run stanc3 --print-canonical to update
+       print an updated version of your code."
+    3. A timeline which includes a date or version after which this feature is
+       slated for removal, for example "This deprecation will expire in Stan
+       version 2.31.0 (expected Sept. 2022)".
+2. It has been added to the [documentation on
+   deprecation](https://mc-stan.org/docs/reference-manual/deprecated-features-appendix.html)
+   with the same 3 pieces of information, and additional information as needed
+   (such as any of: reasoning for removal, advantages of new behavior, ongoing
+   related changes, etc).
+3. Examples in the documentation have been updated to remove any examples of the
+   deprecated syntax, except for those referring to its deprecation. This change
+   may lag behind bullets 1 and 2, but must occur before the removal.
```

Comment:
The main documentation should remove a feature and include it in the deprecation section as soon as the feature is deprecated.   Otherwise, users will still think it's fine.

Reply:
I agree, I mainly mean that updating each example in all other parts of the dics may not happen right away. We could require this if we wanted, and it may even serve as a reasonable barrier to spurious deprecations. I didn't want to make it blocking like the first two bullets, mainly because it is a much larger task

Reply:
I'd like to keep the docs up to date, so we should require the doc fix when the deprecations go in.  I'm always happy to help revising doc.